packages/@tailwindcss-upgrade/src/template/codemods/modernize-arbitrary-values.test.ts

```diff
@@ -0,0 +1,71 @@
+import { __unstable__loadDesignSystem } from '@tailwindcss/node'
+import { expect, test } from 'vitest'
+import { modernizeArbitraryValues } from './modernize-arbitrary-values'
+
+test.each([
+  // Arbitrary variants
+  ['[[data-visible]]:flex', 'data-visible:flex'],
+  ['[&[data-visible]]:flex', 'data-visible:flex'],
+  ['[[data-visible]&]:flex', 'data-visible:flex'],
+  ['[&>[data-visible]]:flex', '*:data-visible:flex'],
+  ['[&_>_[data-visible]]:flex', '*:data-visible:flex'],
+
+  ['[&_[data-visible]]:flex', '**:data-visible:flex'],
+
+  ['[&:first-child]:flex', 'first:flex'],
+  ['[&:not(:first-child)]:flex', 'not-first:flex'],
+
+  // nth-child
+  ['[&:nth-child(2)]:flex', 'nth-2:flex'],
+  ['[&:not(:nth-child(2))]:flex', 'not-nth-2:flex'],
+
+  ['[&:nth-child(-n+3)]:flex', 'nth-[-n+3]:flex'],
+  ['[&:not(:nth-child(-n+3))]:flex', 'not-nth-[-n+3]:flex'],
+  ['[&:nth-child(-n_+_3)]:flex', 'nth-[-n+3]:flex'],
+  ['[&:not(:nth-child(-n_+_3))]:flex', 'not-nth-[-n+3]:flex'],
+
+  // nth-last-child
+  ['[&:nth-last-child(2)]:flex', 'nth-last-2:flex'],
+  ['[&:not(:nth-last-child(2))]:flex', 'not-nth-last-2:flex'],
+
+  ['[&:nth-last-child(-n+3)]:flex', 'nth-last-[-n+3]:flex'],
+  ['[&:not(:nth-last-child(-n+3))]:flex', 'not-nth-last-[-n+3]:flex'],
+  ['[&:nth-last-child(-n_+_3)]:flex', 'nth-last-[-n+3]:flex'],
+  ['[&:not(:nth-last-child(-n_+_3))]:flex', 'not-nth-last-[-n+3]:flex'],
+
+  // nth-child odd/even
+  ['[&:nth-child(odd)]:flex', 'odd:flex'],
+  ['[&:not(:nth-child(odd))]:flex', 'even:flex'],
+  ['[&:nth-child(even)]:flex', 'even:flex'],
+  ['[&:not(:nth-child(even))]:flex', 'odd:flex'],
+
+  // Keep multiple attribute selectors as-is
+  ['[[data-visible][data-dark]]:flex', '[[data-visible][data-dark]]:flex'],
+
+  // Complex attribute selectors with operators, quotes and insensitivity flags
+  ['[[data-url*="example"]]:flex', 'data-[url*="example"]:flex'],
+  ['[[data-url$=".com"_i]]:flex', 'data-[url$=".com"_i]:flex'],
+  ['[[data-url$=.com_i]]:flex', 'data-[url$=.com_i]:flex'],
+
+  // Attribute selector wrapped in `&:is(…)`
+  ['[&:is([data-visible])]:flex', 'data-visible:flex'],
+
+  // Compound arbitrary variants
+  ['has-[[data-visible]]:flex', 'has-data-visible:flex'],
+  ['has-[&:is([data-visible])]:flex', 'has-data-visible:flex'],
```

Comment:
`has-[&:is([data-visible])]:flex` has a specificity of `0, 3, 0` in v3 but `[has-[&:is([data-visible])]:flex](has-data-visible:flex)` is `0, 2, 0` in v4. 


HOWEVER `has-[&:is([data-visible])]:flex` is also `0, 2, 0` in v4 so keeping this wouldn't fix it either.

Reply:
Yeah exactly, it's a side effect of the implementation in v4 unrelated to this change. If you want to be 100% the same, we should switch to arbitrary variants instead but that makes it worse.